File: Interview Problems/Parking Lot/Java Code/controller/EntryController.java
package controller;

import domain.ParkingSlot;
import domain.Ticket;
import domain.Vehicle;
import repository.VehicleRepository;
import service.SlotService;
import service.TicketService;

import java.util.Optional;

public class EntryController {

    private final SlotService slotService;
    private final TicketService ticketService;
    private final VehicleRepository vehicleRepository;

    public EntryController(SlotService slotService,
                           TicketService ticketService,
                           VehicleRepository vehicleRepository) {
        this.slotService = slotService;
        this.ticketService = ticketService;
        this.vehicleRepository = vehicleRepository;
        System.out.println("[CONTROLLER] EntryController initialized");
    }

    public EntryResult enterVehicle(String licensePlate, Vehicle.VehicleType vehicleType) {
        System.out.println("[CONTROLLER] Vehicle entry request - Plate: " + licensePlate
                + " | Type: " + vehicleType);
        try {
            // Reuse existing vehicle record or register a new one
            Vehicle vehicle = vehicleRepository.findByLicensePlate(licensePlate)
                    .orElseGet(() -> {
                        Vehicle v = new Vehicle(licensePlate, vehicleType);
                        vehicleRepository.save(v);
                        return v;
                    });

            // Allocate slot (throws if none available)
            ParkingSlot slot = slotService.allocateSlot(vehicleType);

            // Generate and persist ticket
            Ticket ticket = ticketService.generateTicket(vehicle, slot);

            System.out.println("[CONTROLLER] Vehicle entry successful - Ticket: " + ticket.getId());
            return new EntryResult(true, ticket, "Entry successful");

        } catch (Exception e) {
            System.out.println("[CONTROLLER] Vehicle entry failed: " + e.getMessage());
            return new EntryResult(false, null, e.getMessage());
        }
    }

    // ─── Inner DTO ────────────────────────────────────────────────────────────

    public static class EntryResult {
        private final boolean success;
        private final Ticket ticket;
        private final String message;

        public EntryResult(boolean success, Ticket ticket, String message) {
            this.success = success;
            this.ticket = ticket;
            this.message = message;
        }

        public boolean isSuccess() { return success; }
        public Ticket getTicket() { return ticket; }
        public String getMessage() { return message; }

        @Override
        public String toString() {
            return "EntryResult{success=" + success
                    + ", ticketId=" + (ticket != null ? ticket.getId() : "null")
                    + ", message=" + message + "}";
        }
    }
}

File: Interview Problems/Parking Lot/Python Code/controller/entry_controller.py
from typing import Optional
from domain.vehicle import Vehicle, VehicleType
from domain.ticket import Ticket
from repository.vehicle_repository import VehicleRepository
from service.slot_service import SlotService
from service.ticket_service import TicketService

class EntryResult:
    def __init__(self, success: bool, ticket: Optional[Ticket], message: str):
        self.success = success
        self.ticket = ticket
        self.message = message

    def __repr__(self):
        ticket_id = self.ticket.id if self.ticket else "null"
        return f"EntryResult(success={self.success}, ticketId={ticket_id}, message='{self.message}')"

class EntryController:
    def __init__(self, slot_service: SlotService, ticket_service: TicketService, vehicle_repo: VehicleRepository):
        self.slot_service = slot_service
        self.ticket_service = ticket_service
        self.vehicle_repository = vehicle_repo
        print("[CONTROLLER] EntryController initialized")

    def enter_vehicle(self, license_plate: str, vehicle_type: VehicleType) -> EntryResult:
        print(f"[CONTROLLER] Vehicle entry request - Plate: {license_plate} | Type: {vehicle_type.name}")
        try:
            vehicle = self.vehicle_repository.find_by_license_plate(license_plate)
            if not vehicle:
                vehicle = Vehicle(license_plate, vehicle_type)
                self.vehicle_repository.save(vehicle)

            slot = self.slot_service.allocate_slot(vehicle_type)
            ticket = self.ticket_service.generate_ticket(vehicle, slot)

            print(f"[CONTROLLER] Vehicle entry successful - Ticket: {ticket.id}")
            return EntryResult(True, ticket, "Entry successful")

        except Exception as e:
            print(f"[CONTROLLER] Vehicle entry failed: {str(e)}")
            return EntryResult(False, None, str(e))
